Refactor tool-scripts to use one script

portante opened the pull request:
Addresses issue #1459 following "Proposal 1".

This is a fairly invasive and major change to the tool sub-system.  The
pbench-agent no longer tries to install **ANY** RPMs via `yum`.  The
interface to the `pbench-start/stop/postprocess-tools` has changed, and
changes to various interfaces and behaviors of individual tools have
also been made.

**Behavioral Changes**

The pbench-agent no longer tries to install *ANY* RPMs via `yum`.

This is significant in that all of the benchmark scripts now only
*check* that their required RPMs are installed, and all the tools
during registration only perform a check that their requirements are
satisfied.

Again, no RPMs are installed by the pbench-agent code.

All of the tool datalogs have been enhanced to check for their required
environment and issue error messages describing what is missing for them
to operate properly.

One major behavior change to the `systemtap` tool is worth noting.
The systemtap script argument is now the "compiled" systemtap
script. The user is now responsible for making sure the provided
script is available to the systemtap tool on all the hosts it is
registered.

**Interface Changes**

We are removing knowledge of how `bench-scripts` handle "iterations" and
"samples" so that the tools are only given a directory to work with,
`${tool_output_dir}`, and no longer use `${pbench_tmp}`.  All that they
need to do with files are managed in the `${tool_output_dir}`.

This then allowed us to remove the `--iteration` and `--group`
parameters from all the tools, and in turn, remove the `--iteration`
parameter from the `pbench-(start|stop|postprocess)-tools` commands.
This prevents knowledge of how benchmark scripts work from leaking into
the various tool implementations.

**Tool Script Changes**

We now have two scripts that implement the methods for running tools:

 * `base-tool`
    * a merger of all the previous tool script interfaces
    * with PR #1248, this will be replaced by Tool Meister code
    * all tools now provide a proper `${tool}-datalog` implementation,
      along with `${tool}-postprocess` and `${tool}-stop-postprocess`
      scripts where appropriate
    * the `haproxy-ocp`, `oc`, and `pprof` tools now have a Python 3
      based `${tool}-datalog`
 * `external-data-source`
    * dedicated to just handling for the no-op external data source

The `user-tool` interfaces just provide an arbitrary path to scripts
that should behave the same way as our existing `*-datalog`,
`*-stop-postprocess`, and `*-postprocess` scripts.  The `user-tool`
start script is passed the `${tool_output_dir}` and the interval
value.  The stop and post-process scripts are just passed the
`${tool_output_dir}`.

**Tool Post-Processing**

We have added the notion of a "stop-postprocess" script so that when
tools are stopped (i.e. when `pbench-stop-tools` is invoked), any
post-processing required to be performed on the host on which the
tool is running can be executed.  This allows us to run all the
normal post-procesing of data to generate charts and graphs during
the `pbench-postprocess-tools` phase only on the local controller, or
at a later time.

**Unit Tests**

We have added unit tests for all the tools in the `tool-scripts`
directory, and in the `tool-scripts/datalog` directory.

**A Guide to Reviewing the Changes**

 * `agent/base`

   Changed to only check for RPMs, installation removed; we also now
   have a single `_PBENCH_UNIT_TESTS` environment variable in use

 * `agent/util-scripts`

   Changes to remove the use of `--iteration` when invoke tools;
   accommodate new list of available tools

 * `agent/bench-scripts`

   Changes to remove the use of `--iteration` when invoking tools

 * `agent/tool-scripts`

   Except for the "external data source", all 37 tool files are now
   symlinks to `base-tool`, with it containing all the tool iterface
   code and how the `${tool}-datalog`, `${tool}-stop-postprocess`,
   `${tool}-postprocess` are invoked

 * `agent/tool-scripts/datalog`

   All tools now have a `${tool}-datalog`; the `pidstat-convert` script
   now lives in this directory, moved from `../postprocess`; we share
   a common `File-Capture-datalog` for `proc-interrupts`,
   `proc-sched_debug`, and `proc-vmstat`

 * `agent/tool-scripts/postprocess`

   No-op post-processing scripts have been removed (`dm-cache`,
   `systemtap`, `tcpdump`, and `turbostat`); we have added
   `${tool}-stop-postprocess` scripts where necessary (`blktrace`,
   `cpuacct`, `kvmtrace`, `mpstat`, and `perf`); a number of the
   post-processing scripts have been fixed to accommodate the changes to
   their `-datalog` counter parts.

portante commented:
@ndokos, asked you to review because this now has passing unit tests on the agent side, and except for the five tools that need refactoring, the majority of the work is done.

riya-17 commented:
What is the usage of mock-bin?

zulcss commented on agent/tool-scripts/datalog/prometheus-metrics-datalog:
```diff
@@ -1,35 +1,165 @@
-            key, host, port)
+    # First, we copy the inventory file provided to our tool output directory
+    # for safe keeping.
+    import shutil
```
Lets group the imports all near the top

portante replied:
Is there a reason to do that?  By only importing where you need it, we avoid importing too many modules when code paths don't end up using them.

zulcss replied:
https://www.python.org/dev/peps/pep-0008/#imports

portante replied:
Let's resolve this when we have the `flake8` and `black` enforcing in place.

zulcss commented on agent/tool-scripts/datalog/prometheus-metrics-datalog:
```diff
@@ -1,35 +1,165 @@
+import math
+import subprocess
+from time import time, sleep
+
```
ditto

zulcss commented on agent/tool-scripts/postprocess/prometheus-metrics-postprocess:
```diff
@@ -1,12 +1,25 @@
-input_file = sys.argv[1]
-output_file = sys.argv[2]
-docs = []
+import json
```
Move to line 6

zulcss commented on agent/tool-scripts/postprocess/prometheus-metrics-postprocess:
```diff
@@ -44,7 +57,7 @@ def convert_value(val):
 
 # denormalize metrics for elasticsearch to scale
-def denormalize_metrics(new_item, metrics):
+def denormalize_metrics(new_item, metrics, docs):
```
Comment should go before line 61

zulcss commented on agent/tool-scripts/datalog/oc-datalog:
```diff
@@ -0,0 +1,161 @@
+def now():
+    return datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%S.%f")
+
+import subprocess
```
move to before line 4

portante replied:
Will address after `black` and `flake8` are in enforcing mode.

zulcss commented on agent/tool-scripts/datalog/oc-datalog:
```diff
@@ -0,0 +1,161 @@
+            file=sys.stderr)
+    sys.exit(1)
+
+from datetime import datetime
```
move to before line 4

portante replied:
Will address after `black` and `flake8` are in enforcing mode.

portante commented:
@ndokos, @zulcss, I have refactored this PR into two commits, the first being just code changes, the second being all the test infrastructure changes.  I have also updated the PR description to add additional description about the changes.

portante commented on agent/tool-scripts/base-tool:
```diff
@@ -0,0 +1,643 @@
+
+# Defaults specific to certain tools
+case "${tool}" in
+blktrace)
```
No, see below, only a few tools actually check for the presence of a RPM.

portante commented on agent/tool-scripts/base-tool:
```diff
@@ -0,0 +1,643 @@
+		error_log "INTERNAL ERROR: unsupported tool, ${tool}"
+		exit 1
+		;;
+	esac
```
Right, the Tool Meister work, where we can do that in Python 3 is the place for that change.

portante commented on agent/tool-scripts/base-tool:
```diff
@@ -0,0 +1,643 @@
+	tool_stdout_file="${tool_output_dir}/${tool}-stdout.txt"
+	tool_stderr_file="${tool_output_dir}/${tool}-stderr.txt"
+	${tool_cmd_file} > "${tool_stdout_file}" 2> "${tool_stderr_file}" & echo ${!} > "${tool_pid_file}"
+	wait
```
The background tool command that was issued.

portante commented on agent/tool-scripts/datalog/docker-info-datalog:
```diff
@@ -1,17 +1,24 @@
+rc=0
+while [[ ${rc} -eq 0 ]]; do
+	echo "timestamp: $(date +%s.%N)"
+	docker info
```
I thought about doing that, but I am hoping to further that kind of work via Python 3 with the Tool Meister work.

portante commented on agent/tool-scripts/base-tool:
```diff
@@ -0,0 +1,643 @@
+			rc=${?}
+		fi
+		;;
+	esac
```
Yes, no other tools install packages.

natoscott commented on agent/tool-scripts/postprocess/pcp-postprocess:
```diff
@@ -1,50 +1,80 @@
+    exit 1
+fi
+
+command -v pmdumptext > /dev/null
```
@portante the use of pmdumptext here could be replaced with the more modern PCP utility pcp2csv(1) - its a python script, part of the pcp-system-tools package on RPM systems (and core pcp package on DEB boxen).  pmdumptext(1) is from the Qt-dependent, optional pcp-gui package, so you'll likely have fewer dependencies as well as many more formatting options.

portante replied:
I'll be adding a separate commit to this patch set which will show only the changes to use `pcp2csv`.

zulcss commented on agent/tool-scripts/datalog/turbostat-datalog:
```diff
@@ -1,6 +1,23 @@
+	exit 1
+fi
+
+if grep -q KVM /sys/devices/virtual/dmi/id/product_name; then
```
systemd-virt-detect

portante replied:
Seems to be everywhere `systemd` is.  I'll add a patch that shows just that change to use `systemd-detect-virt`.

ndokos commented on agent/util-scripts/pbench-postprocess-tools:
```diff
@@ -1,5 +1,5 @@
 #!/bin/bash
-# -*- mode: shell-script; indent-tabs-mode: t; sh-basic-offset: 8; sh-indentation: 8; sh-indent-for-case-alt: + -*-
+# -*- mode: shell-script; indent-tabs-mode: t; sh-basic-offset: 8; sh-indentation: 8; tab-width: 8 -*-
```
indent-tabs-mode nil? shouldn't offsets and indents be 4?

portante replied:
I am not sure why we want the `indent-tabs-mode` to be `nil` for this file, when all other files are pretty much using `t`.  Also, we are just changing the tab-width to be 8, but the original indentation before this change was already 8, so why do we want to change it to 4 in this commit?

ndokos commented on agent/util-scripts/pbench-postprocess-tools:
```diff
@@ -9,26 +9,24 @@ action=`echo ${script_name#pbench-} | awk -F- '{print $1}'`
 . "$pbench_bin"/base
 
-if [ "$action" != "start" -a "$action" != "stop" -a "$action" != "postprocess" -a "$action" != "kill" ]; then 
+if [ "$action" != "start" -a "$action" != "stop" -a "$action" != "postprocess" -a "$action" != "kill" ]; then
```
Use [[...]] and braces around variable names?

portante replied:
Fixed.

portante commented on agent/tool-scripts/datalog/pidstat-datalog:
```diff
@@ -0,0 +1,60 @@
+default_options="-l -H -w -u -h -d -r ${threads_arg} -p ALL ${patterns_arg}"
+
+if [[ -z "${_tool_bin}" ]]; then
+	_tool_bin="/usr/local/bin/mpstat"
```
This is because we rely on the `pbench-sysstat` RPM we build which installs the sysstat commands under the `/usr/local/bin` path.

portante commented on agent/bench-scripts/pbench-trafficgen:
```diff
@@ -1116,11 +1116,11 @@ function inner_most_loop() {
 						if [ $rc -ne 0 ]; then
 							error_log "iteration $iteration sample $sample returned non-zero exit code"
-							pbench-stop-tools --group=$tool_group --iteration=$iteration --dir=$benchmark_results_dir
+							pbench-stop-tools --group=$tool_group --dir=$benchmark_results_dir
```
Because `trafficgen` uses tool triggers to start/stop tools, but if a problem arises, it manually stops them.

portante commented:
@ndokos, regarding the `cpuacct-datalog` test failures, the latest update should fix them so that the datalog does not actually use the `/sys/fs/cgroup/cpuacct/` tree.  Also, the `disk-datalog` has been fixed to not try to escape the `:`.

As for the Fedora 31 test failures related to perl, is that something we want to fix in this commit along with all the other changes we have?  Or should we consider fixing them in another commit?

ndokos commented:
As long as Travis CI is happy, I think we should leave well enough alone. F31 is just the canary in the mine: we know we are going to have a problem in the future (which is better than finding out that we have a problem *now* ;-) )

ndokos commented on agent/tool-scripts/datalog/haproxy-ocp-datalog:
```diff
@@ -0,0 +1,372 @@
+
+    def wait(self, all=True):
+        # We wait on any "metrics" background process first. Once that
+        # completes, we "reap" (poll()) and other background process so that
```
and --> any?

portante replied:
Fixed.

ndokos commented on agent/tool-scripts/postprocess/pidstat-postprocess:
```diff
@@ -82,34 +81,6 @@ if (scalar @ARGV > 2) {
 		}
 	}
 	close FH;
```
This whole section got deleted, presumably because it should go into pidstat-stop-process, but there is no pidstat-stop-postrpocess in the commit.

portante replied:
Actually, I could not find where `$vcpus` was being used in the original code.  So I just removed it entirely.

ndokos commented on agent/tool-scripts/postprocess/cpuacct-stop-postprocess:
```diff
@@ -0,0 +1,24 @@
+	opendir(D, $sysdevsyscpu) || die "Can't open directory $sysdevsyscpu: $!\n";
+	my @dirlist = readdir(D);
+	closedir(D);
+    open(CPUNODEMAP, ">$dir/$outfile") || die "Could not create $dir/$outfile";
```
White space?

portante replied:
Fixed.

ndokos commented on agent/tool-scripts/postprocess/kvmtrace-stop-postprocess:
```diff
@@ -151,3 +178,7 @@ close(KVMTRACE_EXIT_SOURCES);
+
+waitpid($zip_kallsyms_pid, 0);
+waitpid($zip_report_pid, 0);
+waitpid($zip_trace_pid, );
```
This looks as if it would be better written in bash: It does exec/system mostly. What does perl buy us?

portante replied:
Not sure what you mean.  Isn't this whole file written in `perl` already?  Not sure this commit wants to take on re-writing this code as well.

ndokos commented on agent/tool-scripts/postprocess/cpuacct-postprocess:
```diff
@@ -82,11 +92,7 @@ while (my $line = <CPUACCT_TXT>) {
-			} else {
-				$node_id = "0";
-			}
+            $node_id = cpu_node($tgt);
```
White space?

portante replied:
Fixed.

portante commented on agent/tool-scripts/datalog/oc-datalog:
```diff
@@ -0,0 +1,161 @@
+
+class GracefulPopen(subprocess.Popen):
+    """Simple sub-class of Popen which adds a graceful way to terminate a
+    process.
```
Will address after `black` and `flake8` are in enforcing mode.

portante commented on agent/tool-scripts/datalog/oc-datalog:
```diff
@@ -0,0 +1,161 @@
+
+# Establish signal handler for TERM, QUIT, and INT to stop process creation,
+# and then tear them all down.
+import signal
```
Will address after `black` and `flake8` are in enforcing mode.

portante commented on agent/tool-scripts/datalog/oc-datalog:
```diff
@@ -0,0 +1,161 @@
+signal.signal(signal.SIGQUIT, handler)
+signal.signal(signal.SIGINT, handler)
+
+import time
```
Will address after `black` and `flake8` are in enforcing mode.

portante commented on agent/tool-scripts/datalog/pprof-datalog:
```diff
@@ -0,0 +1,354 @@
+else:
+    # First, we copy the inventory file provided to our tool output directory
+    # for safe keeping.
+    import shutil
```
Will address after `black` and `flake8` are in enforcing mode.

portante commented on agent/tool-scripts/datalog/pprof-datalog:
```diff
@@ -0,0 +1,356 @@
+    from ansible.inventory.manager import InventoryManager
+except ImportError as exc:
+    logging.error("[%s]: required Ansible python3 modules not available, '%s'",
+            PROG, exc)
```
Will address after `black` and `flake8` are in enforcing mode.

portante commented on agent/tool-scripts/datalog/pprof-datalog:
```diff
@@ -0,0 +1,356 @@
+import math
+import errno
+from time import time, sleep
+from datetime import datetime
```
Will address after `black` and `flake8` are in enforcing mode.

portante commented on agent/util-scripts/require-rpm:
```diff
@@ -0,0 +1,20 @@
+    _version=""
+fi
+if [[ ! -z "${_PBENCH_UNIT_TESTS}" ]]; then
+    # don't install packages when unit-testing
```
Fixed.

portante commented on agent/bench-scripts/pbench-dbench:
```diff
@@ -400,7 +400,7 @@ for thread_count in `echo $thread_counts | sed -e s/,/" "/g`; do
 							final_cmd_file="ssh $ssh_opts $client \"$benchmark_client_cmd_file\""
 						fi
-						if [[ -z "$_PBENCH_BENCH_TESTS" ]]; then
+						if [[ -z "$_PBENCH_UNIT_TESTS" ]]; then
```
Fixed this instance.

portante commented on agent/tool-scripts/base-tool:
```diff
@@ -0,0 +1,858 @@
+		;;
+	systemtap)
+		# The installation for systemtap includes compiling the kernel
+		# module from the stap script; this is done to avoid extra
```
Fixed.

portante commented on agent/tool-scripts/datalog/jstack-datalog:
```diff
@@ -0,0 +1,75 @@
+	fi
+	jstack_stdout_file=${tool_output_dir}/jstack-${pname}-stdout.txt
+
+	# check if containerized.
```
Not sure how we can use `systemd-detect-virt` to see if another PID is containerized.

portante commented on agent/tool-scripts/datalog/jstack-datalog:
```diff
@@ -0,0 +1,75 @@
+#!/bin/bash
+# -*- mode: shell-script; indent-tabs-mode: t; sh-basic-offset: 8; sh-indentation: 8; tab-width: 8 -*-
```
Fixed.

portante commented on agent/tool-scripts/datalog/jmap-datalog:
```diff
@@ -0,0 +1,63 @@
+    for jvm_pid in ${jvm_pids}; do
+	# TODO: extend the heap stats to collect min,max and non-heap
+	# check if containerized.
+	cnt=$(grep -cF 'user.slice' /proc/${jvm_pid}/cgroup 2> /dev/null)
```
Fixed.

portante commented on agent/tool-scripts/datalog/jmap-datalog:
```diff
@@ -0,0 +1,63 @@
+#!/bin/bash
+# -*- mode: shell-script; indent-tabs-mode: t; sh-basic-offset: 8; sh-indentation: 8; tab-width: 8 -*-
```
Fixed.

portante commented on agent/util-scripts/require-rpm:
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
+# -*- mode: shell-script; indent-tabs-mode: t; sh-basic-offset: 4; sh-indentation: 4; tab-width: 8 -*-
```
Why `nil`?

portante commented on agent/base:
```diff
@@ -1,4 +1,5 @@
 #!/bin/bash
+# -*- mode: shell-script; indent-tabs-mode: t; sh-basic-offset: 4; sh-indentation: 4; tab-width: 8 -*-
```
Not sure why we want `nil`?  It seems to be `t` for almost every other file.

portante commented:
@natoscott, can you take a look at the "Update to use pcp2csv" commit?

portante commented:
@ndokos, I have addressed various code-review comments in the "Address code review comments" commit to see what has been asked for isolated from the other commits.

@zulcss, please let me know if the "Use systemd-detect-virt in turbostat datalog" looks good.

I'd like to get this merged ASAP.

natoscott commented:
> + ./cpuacct-datalog: line 12: pushd: /sys/fs/cgroup/cpuacct: No such file or directory

This is almost certainly due to the systemd transition to using the kernel cgroup2 subsystem exclusively in f31 (i.e. it no longer starts up in 'unified' mode by default anymore).

ndokos commented:
I thought you wanted to squash this to one commit before merging. What happened?

portante commented:
Bummer, yes, took quick on the draw, I'll revert and redo this.
